Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -116,271 +53,113 @@ COREGRAPHICS_EXPORT void CGImageAddDestructionListener(CGImageDestructionListene
 */
-CGImageRef CGImageCreateWithImageInRect(CGImageRef ref, CGRect rect) {
-    if (ref == NULL) {
-        TraceWarning(TAG, L"CGImageCreateWithImageInRect: ref = NULL!");
-        return 0;
-    }
-
-    rect = CGRectIntegral(rect);
+CGImageRef CGImageCreate(size_t width,
+                         size_t height,
+                         size_t bitsPerComponent,
+                         size_t bitsPerPixel,
+                         size_t bytesPerRow,
+                         CGColorSpaceRef colorSpace,
+                         CGBitmapInfo bitmapInfo,
+                         CGDataProviderRef provider,
+                         const float* decode,
+                         bool shouldInterpolate,
+                         CGColorRenderingIntent intent) {
+    RETURN_NULL_IF(((provider == nullptr) || ![(NSObject*)provider isKindOfClass:[NSData class]]));
 
-    CGRect imgRefSize;
-    imgRefSize.origin.x = 0;
-    imgRefSize.origin.y = 0;
-    imgRefSize.size.width = (float)ref->Backing()->Width();
-    imgRefSize.size.height = (float)ref->Backing()->Height();
+    NSData* dataProvider = (__bridge NSData*)provider;
 
-    rect = CGRectIntersection(rect, imgRefSize);
+    unsigned char* data = (unsigned char*)[dataProvider bytes];
 
-    __CGSurfaceInfo surfaceInfo;
-    ref->Backing()->GetSurfaceInfoWithoutPixelPtr(&surfaceInfo);
+    bool colorSpaceAllocated = false;
+    if (colorSpace == NULL) {
+        if (bytesPerRow >= (width * 3)) {
+            TraceWarning(TAG, L"Warning: colorSpace = NULL, assuming RGB based on bytesPerRow.");
+            colorSpace = CGColorSpaceCreateDeviceRGB();
+        } else {
+            TraceWarning(TAG, L"Warning: colorSpace = NULL, assuming Gray based on bytesPerRow.");
+            colorSpace = CGColorSpaceCreateDeviceGray();
+        }
 
-    // Override width and height with the rect
-    surfaceInfo.width = rect.size.width;
-    surfaceInfo.height = rect.size.height;
+        colorSpaceAllocated = true;
+    }
 
-    assert(surfaceInfo.surfaceData == NULL);
+    ComPtr<IWICBitmap> image;
+    ComPtr<IWICImagingFactory> imageFactory = _GetWICFactory();
 
-    CGImageRef newImage = new CGBitmapImage(surfaceInfo);
+    REFGUID pixelFormat = _CGImageGetWICPixelFormat(bitsPerComponent, bitsPerPixel, colorSpace, bitmapInfo);
 
-    int startX = (int)rect.origin.x;
-    int startY = (int)rect.origin.y;
-    int sizeX = (int)rect.size.width;
-    int sizeY = (int)rect.size.height;
+    HRESULT status = imageFactory->CreateBitmapFromMemory(width, height, pixelFormat, bytesPerRow, height * bytesPerRow, data, &image);
 
-    BYTE* srcIn =
-        ((BYTE*)ref->Backing()->LockImageData()) + startY * ref->Backing()->BytesPerRow() + startX * ref->Backing()->BytesPerPixel();
-    BYTE* destOut = (BYTE*)newImage->Backing()->LockImageData();
+    CGImageRef imageRef = __CGImage::CreateInstance();
+    imageRef->ImageSource().Attach(image.Detach());
```

Comment:
.Attach with .Detach is generally wrong. ComPtr Attach and Detach should only be used when going to and from a raw */** which should usually be happening only at ABI boundaries / to interop with other APIs that expect void* or other bad patterns that we can't change. 

Without reading more here it seems ImageSource() is return value (r val) to which the assignment would be lost in the next line as no one is holding on to it. It would be "clever" of you to be returning a ComPtr& from ImageSource() as a mechanism around this but that breaks encapsulation as bad a public member, introduces code like this that appears incorrect at first glance, and is arguable just as much or more typing than a traditional setter. In general try to ask yourself if being "clever" is really necessary whenever you feel like you need to be.